pkg/logs/console.go

```diff
@@ -47,9 +47,11 @@ var colors = []brush{
 
 // consoleWriter implements LoggerInterface and writes messages to terminal.
 type consoleWriter struct {
-	lg       *logWriter
-	Level    int  `json:"level"`
-	Colorful bool `json:"color"` //this filed is useful only when system's terminal supports color
+	lg                 *logWriter
+	UseCustomFormatter bool
```

Comment:
we don't need to use those two fields. We could do this like:

```go
type consoleWriter struct {
     lg *logWriter
     formatter LogFormatter // use the interface defined in your last PR
    Formatter string `json:"formatter"`
    Level int
    Colorful bool
}

func (c *consoleWriter) Init(jsonConfig string) error {
	if len(jsonConfig) == 0 {
		return nil
	}
        res := json.Unmarshal([]byte(jsonConfig), c)
        if res.Formatter != "" {
                // we need two more methods, RegisterLogFormatter(name string, formatter LogFormatter) and 
                // GetFormatter(name string)
                res.formatter = GetFormatter(res.Formatter)
        } else {
                // In your last PR, the consoleWriter already implement LogFormatter interface.
                // if user do not use custom formatter, we use consoleWriter default formatter(itself)
                res.formatter = res
        }
	return res
}

// WriteMsg writes message in console.
func (c *consoleWriter) WriteMsg(lm *LogMsg) error {
	msg := c.formatter.Format(lm)
	c.lg.writeln(msg)
	return nil
}
```
Other implementations are similar to this.

Reply:
One problem that I'm thinking about that if every adapter has a `LogFormatter` field it means it must have a `Format` method. This isn't an issue when you don't want to use custom formatting. In my last PR I hardcoded in a format method for each adapter like so:
```go
func (c *consoleWriter) Format(lm *LogMsg) string {
	msg := lm.Msg

	if c.Colorful {
		msg = strings.Replace(lm.Msg, levelPrefix[lm.Level], colors[lm.Level](levelPrefix[lm.Level]), 1)
	}

	h, _, _ := formatTimeHeader(lm.When)
	bytes := append(append(h, msg...), '\n')

	return string(bytes)

}
```
In order to get a custom formatter this `Format` function would need to be overrided somehow with the user defined logging formatter and I'm not sure how to do that.

I also think that putting the formatter in as its own parameter is better as it seperates the logic better
If the formatter was part of the config string it would be something like
```go
logs.SetLoggerWithOpts("console", []string{"something":"key","formatter":"customFormatter"})
// instead of 
logs.SetLoggerWithOpts("console", []string{""}, customFormatter)
```
I also don't know how you'd put a function into a JSON string then extract the a pointer to the function when you decode the JSON config in the `init` function

Reply:
> One problem that I'm thinking about that if every adapter has a LogFormatter field it means it must have a Format method. This isn't an issue when you don't want to use custom formatting. In my last PR I hardcoded in a format method for each adapter like so:

My idea is that:
```go
func (c *consoleWriter) Format(lm *LogMsg) string {
	msg := lm.Msg

	if c.Colorful {
		msg = strings.Replace(lm.Msg, levelPrefix[lm.Level], colors[lm.Level](levelPrefix[lm.Level]), 1)
	}

	h, _, _ := formatTimeHeader(lm.When)
	bytes := append(append(h, msg...), '\n')

	return string(bytes)

}
```

Making this method as the default formatter for consoleWritter.
I prefer to define a new default formatter for consoleWritter and remove `Format` method from `Logger` interface.

```
type consoleWritterDefaultFormatter {
     colorful bool
}


func (c *consoleWritterDefaultFormatter) Format(lm *LogMsg) string {
	msg := lm.Msg

	if c.colorful{
		msg = strings.Replace(lm.Msg, levelPrefix[lm.Level], colors[lm.Level](levelPrefix[lm.Level]), 1)
	}

	h, _, _ := formatTimeHeader(lm.When)
	bytes := append(append(h, msg...), '\n')

	return string(bytes)
}

type consoleWriter struct {
     lg *logWriter
     formatter LogFormatter // use the interface defined in your last PR
    Level int
    Colorful bool
}

// using "opts", so that not only LogFormatter but also other parameter could be passed.
// Orm use common.Kv as options. But it's not golang typical practice. 
func (c *consoleWriter) Init(jsonConfig string, opts...Option) error {
	if len(jsonConfig) == 0 {
		return nil
	}
        res := json.Unmarshal([]byte(jsonConfig), c)
        formatter := findFormatter(opts)
        if formatter != nil  {
                res.formatter = formatter
        } else {
                res.formatter = &consoleWritterDefaultFormatter{
                      colorful: res.Colorful
                }
        }
	return res
}

// WriteMsg writes message in console.
func (c *consoleWriter) WriteMsg(lm *LogMsg) error {
	msg := c.formatter.Format(lm)
	c.lg.writeln(msg)
	return nil
}

```

The idea is that we use `LogFormatter` too.

Reply:
Ok this is quite similar to my 2nd implementation. I'll get to work on it